pkg/p2p/p2p.go

```diff
@@ -17,9 +17,6 @@ import (
 // Service provides methods to handle p2p Peers and Protocols.
 type Service interface {
 	AddProtocol(ProtocolSpec) error
-	// ConnectNotify connects to the given multiaddress and notifies the topology once the
-	// peer has been successfully connected.
-	ConnectNotify(ctx context.Context, addr ma.Multiaddr) (address *bzz.Address, err error)
```

Comment:
Thank you @pradovic for taking time to clean this method. This change makes this interface more clear. 👍

Reply:
No problem. I think there is still some problems with this kind of architecture in which there are notifiers on the incoming connections but not on the outgoing connections:
1. Now the caller, in this case `debugapi` needs to notify the appropriate components
2. If there are both calls present, as it was, then this is only partially solved. It is solved from the debug_api buyt not in the case of callers of `p2p.Connect`.
3. In both  cases of the second argument, the caller of connect actually needs to know all the notifiers set on the `p2p` in order to mimic the same behavior of the incoming connection. This leads to a potentially problems in the case of multiple notifiers
4. Some duplicated code, as it can be seen in `kademlia` service which is a bit tricky to follow and maintain.

To solve this problem I would propose to have a more granular notifiers, incoming, ougoing on which the services can register + the init method in protocolSpec.

@acud, @janos LMNWYT :)

Reply:
I support the Init method for ProtoclSpec, and also the Close/OnDisconnect/Teardown (not sure about the name). A feature like this is already required for other features, like https://github.com/ethersphere/bee/pull/585#issuecomment-682599786.

Reply:
> 2. If there are both calls present, as it was, then this is only partially solved. It is solved from the debug_api buyt not in the case of callers of p2p.Connect.
3. In both cases of the second argument, the caller of connect actually needs to know all the notifiers set on the p2p in order to mimic the same behavior of the incoming connection. This leads to a potentially problems in the case of multiple notifiers

I'm not sure if I'm following you here. what does these two points mean?

> To solve this problem I would propose to have a more granular notifiers, incoming, outgoing on which the services can register + the init method in protocolSpec.

Can you please explain what you mean by `more granular notifiers`?  an example with the interface you intend to expose might be useful

Reply:
1 & 2
```
// node.go
p2p.SetNotifier(notifier1)
p2p.SetNotifier(notifier2)
p2p.SetNotifier(notifier3)
// some component
p2p.Connect()
notifier1.Connected()
notifier2.Connected()
// oops forgot notifier3
// oops all notifier need to have exposed connected method
```

3.
```
type Interface struct {
OnConnectedIn
OnConnectedOut
OnDisconnect
}

type ProtocolSpec struct {
// what we already have in p2p.go
Init() // should either be both OnConnectedIn & OnConnectedOut  or a bit more sophisticated
}
```

4. Async idempotent wonderland
```
type Interface struct {
OnConnected
OnDisconnected
} 

// protocols are written in the way that they threat both inbound & outbound connections the same
// when registering protocols, we can pay attention to the order in which they are registered and call handlers in the same order
// different types of errors could indicate if the loop should be broken (and disconnected for example) or just continue // disregarding  that error happen in one particular protocol

```

Reply:
about `notifyPeerSig` - this is totally subjective, also `k.connectPeers.Add` is called from different places so maybe we should introduce an abstraction for that too? I'm not sure what are your arguments here. You are free to refactor the package as you wish (although it is questionable whether we should pour engineering hours over something that works).

Reply:
In `idempotent wonderland` example, this would be only on one place probably + I already mentioned before that it's hard to tackle a problem with having abstractions in sync (peer registry & kademlia), so that I would actually rely on peer registry, but let's not get back there :) I will merge the PR :)